### DISCUSSION TITLE
What is the "right" way to add surface stress to 2D model?

### DISCUSSION THREAD

**Discussion by lgloege** [CliMA/Oceananigans.jl#4014] [General]
I am just learning Oceananigans.jl and Julia. To get acquainted with the framework i'm tryig to advect around a passive tracer on a 2D field. Basically integrate forward the advection diffusion equation. I tried modifying the wind stress example in the docs to work with a (Periodic, Periodic, Flat) topology (i.e. a torus). But I am learning you can't apply a `FluxBoundaryCondition` in the `Flat` direction. Any guidance would be appreciated. Thanks! Here is my code so far:

```julia
using Random
using Printf
using TaylorSeries
using CairoMakie

using Oceananigans
using Oceananigans.Units: minute, minutes, hour, hours


architecture = CPU()

Nx = Ny = 128     # number of points in each of horizontal directions
Lx = Ly = 32     # (m) domain horizontal extents

grid = RectilinearGrid(
    architecture,
    topology = (Periodic, Periodic, Flat),
    size = (Nx, Ny),
    x = (0, Lx),
    y = (0, Ly)
)

# seawater buoyancy
buoyancy = SeawaterBuoyancy(
    equation_of_state = LinearEquationOfState(
        thermal_expansion = 2e-4,
        haline_contraction = 8e-4)
)

# Constants (you can adjust these values)
Q = 200.0  # W m⁻², surface heat flux
ρₒ = 1026.0  # kg m⁻³, average density at the surface of the ocean
ρₐ = 1.225  # kg m⁻³, average density of air at sea-level
cᴾ = 3991.0  # J K⁻¹ kg⁻¹, typical heat capacity for seawater

cᴰ = 2.5e-3  # dimensionless drag coefficient
u₁₀ = 10  # m s⁻¹, average wind velocity 10 meters above the ocean

# Wind stress formula (constant wind stress for simplicity)
τx = - ρₐ / ρₒ * cᴰ * u₁₀ * abs(u₁₀)  # m² s⁻²

T_bcs = FieldBoundaryConditions()
S_bcs = FieldBoundaryConditions()
c_bcs = FieldBoundaryConditions()
u_bcs = FieldBoundaryConditions(top = FluxBoundaryCondition(τx))

# setup the model
model = NonhydrostaticModel(; grid, buoyancy,
    advection = UpwindBiased(order=5),
    tracers = (:T, :S, :c),
    coriolis = FPlane(f=1e-4),
    closure = AnisotropicMinimumDissipation(),
    boundary_conditions = (u=u_bcs, T=T_bcs, S=S_bcs, c=c_bcs),
)

# Get the center indices (x, y) 
i_center = Int(grid.Nx / 2)  # Center index in the x-direction
j_center = Int(grid.Ny / 2)  # Center index in the y-direction

# Velocity initial condition: random noise scaled by the friction velocity.
uᵢ(x, y) = sqrt(abs(τx)) #* 1e-3

# Create the passive tracer and initialize it with some value at central grid cell
c_initial = model.tracers.c  
center_x, center_y = Int(grid.Nx / 2), Int(grid.Ny / 2)
c_initial[center_x, center_y] = 1000.0

# `set!` the `model` fields using functions or constants:
set!(model, u=0.0, T=20, S=35, c=c_initial)
```

For what it's worth, here the complete error code:

```
[ Info: Using the advection scheme Upwind Biased reconstruction order 1 in the x-direction because size(grid, 3) = 1
ArgumentError: Cannot set top FluxBoundaryCondition: -0.000298489 in a `Flat` direction!

Stacktrace:
  [1] validate_boundary_condition_topology(bc::BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Float64}, topo::Flat, side::Symbol)
    @ Oceananigans.BoundaryConditions [~/.julia/packages/Oceananigans/ezEr7/src/BoundaryConditions/boundary_condition.jl:130](http://localhost:8889/lab/tree/projects/oceananigans-tutorial/~/.julia/packages/Oceananigans/ezEr7/src/BoundaryConditions/boundary_condition.jl#line=129)
  [2] validate_boundary_conditions(loc::Tuple{DataType, DataType, DataType}, grid::RectilinearGrid{Float64, Periodic, Periodic, Flat, Float64, Float64, Float64, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, Nothing, CPU}, bcs::FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, Nothing, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Float64}, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}})
    @ Oceananigans.Fields [~/.julia/packages/Oceananigans/ezEr7/src/Fields/field.jl:71](http://localhost:8889/lab/tree/projects/oceananigans-tutorial/~/.julia/packages/Oceananigans/ezEr7/src/Fields/field.jl#line=70)
  [3] #apply_regionally!#56
    @ [~/.julia/packages/Oceananigans/ezEr7/src/Utils/multi_region_transformation.jl:121](http://localhost:8889/lab/tree/projects/oceananigans-tutorial/~/.julia/packages/Oceananigans/ezEr7/src/Utils/multi_region_transformation.jl#line=120) [inlined]
  [4] apply_regionally!
    @ [~/.julia/packages/Oceananigans/ezEr7/src/Utils/multi_region_transformation.jl:118](http://localhost:8889/lab/tree/projects/oceananigans-tutorial/~/.julia/packages/Oceananigans/ezEr7/src/Utils/multi_region_transformation.jl#line=117) [inlined]
  [5] macro expansion
    @ [~/.julia/packages/Oceananigans/ezEr7/src/Utils/multi_region_transformation.jl:206](http://localhost:8889/lab/tree/projects/oceananigans-tutorial/~/.julia/packages/Oceananigans/ezEr7/src/Utils/multi_region_transformation.jl#line=205) [inlined]
  [6] Field(loc::Tuple{DataType, DataType, DataType}, grid::RectilinearGrid{Float64, Periodic, Periodic, Flat, Float64, Float64, Float64, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, Nothing, CPU}, data::OffsetArrays.OffsetArray{Float64, 3, Array{Float64, 3}}, bcs::FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, Nothing, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Float64}, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, indices::Tuple{Colon, Colon, Colon}, op::Nothing, status::Nothing)
    @ Oceananigans.Fields ~/.julia/packages/Oceananigans/ezEr7/src/Fields/field.jl:91
  [7] Field(loc::Tuple{DataType, DataType, DataType}, grid::RectilinearGrid{Float64, Periodic, Periodic, Flat, Float64, Float64, Float64, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, Nothing, CPU}, T::DataType; indices::Tuple{Colon, Colon, Colon}, data::OffsetArrays.OffsetArray{Float64, 3, Array{Float64, 3}}, boundary_conditions::FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, Nothing, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Float64}, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, operand::Nothing, status::Nothing)
    @ Oceananigans.Fields ~/.julia/packages/Oceananigans/ezEr7/src/Fields/field.jl:180
  [8] XFaceField
    @ [~/.julia/packages/Oceananigans/ezEr7/src/Fields/field.jl:200](http://localhost:8889/lab/tree/projects/oceananigans-tutorial/~/.julia/packages/Oceananigans/ezEr7/src/Fields/field.jl#line=199) [inlined]
  [9] VelocityFields(grid::RectilinearGrid{Float64, Periodic, Periodic, Flat, Float64, Float64, Float64, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, Nothing, CPU}, user_bcs::@NamedTuple{u::FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, Nothing, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Float64}, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, v::FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, Nothing, Nothing, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, w::FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, Nothing, Nothing, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, T::FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, Nothing, Nothing, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, S::FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, Nothing, Nothing, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, c::FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, Nothing, Nothing, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}})
    @ Oceananigans.Fields [~/.julia/packages/Oceananigans/ezEr7/src/Fields/field_tuples.jl:163](http://localhost:8889/lab/tree/projects/oceananigans-tutorial/~/.julia/packages/Oceananigans/ezEr7/src/Fields/field_tuples.jl#line=162)
 [10] VelocityFields(::Nothing, grid::RectilinearGrid{Float64, Periodic, Periodic, Flat, Float64, Float64, Float64, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, Nothing, CPU}, bcs::@NamedTuple{u::FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, Nothing, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Float64}, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, v::FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, Nothing, Nothing, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, w::FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, Nothing, Nothing, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, T::FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, Nothing, Nothing, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, S::FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, Nothing, Nothing, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, c::FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, Nothing, Nothing, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}})
    @ Oceananigans.Fields [~/.julia/packages/Oceananigans/ezEr7/src/Fields/field_tuples.jl:234](http://localhost:8889/lab/tree/projects/oceananigans-tutorial/~/.julia/packages/Oceananigans/ezEr7/src/Fields/field_tuples.jl#line=233)
 [11] NonhydrostaticModel(; grid::RectilinearGrid{Float64, Periodic, Periodic, Flat, Float64, Float64, Float64, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, Nothing, CPU}, clock::Clock{Float64, Float64}, advection::UpwindBiased{3, Float64, Nothing, Nothing, Nothing, UpwindBiased{2, Float64, Nothing, Nothing, Nothing, UpwindBiased{1, Float64, Nothing, Nothing, Nothing, Nothing, Centered{1, Float64, Nothing, Nothing, Nothing, Nothing}}, Centered{1, Float64, Nothing, Nothing, Nothing, Nothing}}, Centered{2, Float64, Nothing, Nothing, Nothing, Centered{1, Float64, Nothing, Nothing, Nothing, Nothing}}}, buoyancy::SeawaterBuoyancy{Float64, LinearEquationOfState{Float64}, Nothing, Nothing}, coriolis::FPlane{Float64}, stokes_drift::Nothing, forcing::@NamedTuple{}, closure::AnisotropicMinimumDissipation{Oceananigans.TurbulenceClosures.ExplicitTimeDiscretization, Float64, Float64, Nothing}, boundary_conditions::@NamedTuple{u::FieldBoundaryConditions{Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Float64}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}}, T::FieldBoundaryConditions{Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}}, S::FieldBoundaryConditions{Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}}, c::FieldBoundaryConditions{Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Oceananigans.BoundaryConditions.DefaultBoundaryCondition{BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}}}, tracers::Tuple{Symbol, Symbol, Symbol}, timestepper::Symbol, background_fields::@NamedTuple{}, particles::Nothing, biogeochemistry::Nothing, velocities::Nothing, hydrostatic_pressure_anomaly::Oceananigans.Models.NonhydrostaticModels.DefaultHydrostaticPressureAnomaly, nonhydrostatic_pressure::Field{Center, Center, Center, Nothing, RectilinearGrid{Float64, Periodic, Periodic, Flat, Float64, Float64, Float64, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, Nothing, CPU}, Tuple{Colon, Colon, Colon}, OffsetArrays.OffsetArray{Float64, 3, Array{Float64, 3}}, Float64, FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, Nothing, Nothing, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Nothing, Oceananigans.Fields.FieldBoundaryBuffers{Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing}}, diffusivity_fields::Nothing, pressure_solver::Nothing, auxiliary_fields::@NamedTuple{})
    @ Oceananigans.Models.NonhydrostaticModels [~/.julia/packages/Oceananigans/ezEr7/src/Models/NonhydrostaticModels/nonhydrostatic_model.jl:210](http://localhost:8889/lab/tree/projects/oceananigans-tutorial/~/.julia/packages/Oceananigans/ezEr7/src/Models/NonhydrostaticModels/nonhydrostatic_model.jl#line=209)
 [12] top-level scope
    @ In[48]:48
```

**Comment by glwagner** [comment]
2D models cannot have surface stress because there are no z-derivatives in a 2D model.

A surface stress is a vertical flux across the top grid cell. To describe concisely what happens consider a simplified version of the x-momentum equation:

$$ \partial_t u = - \partial_z \tau_z $$

where $\tau$ is the vertical component of the momentum flux. To form a finite volume discretization of this equation we integrate in the vertical. At the top where $k=N$ we have

$$ \partial_t u^N = - \frac{1}{\Delta_z} \left ( \tau_z^\mathrm{top} - \tau_z^N \right ) $$

where $u^N$ is the finite volume representation of $u$ in the surface grid cell (the average of $u$ over that grid cell), $\Delta_z$ is the spacing of the first grid cell, $\tau_z^N$ is the vertical momentum flux across the _first cell interface_ below the surface, and $\tau_z^\mathrm{top}$ is the user-supplied surface boundary condition.

A two-dimensional model has vertical direction and therefore no vertical flux of anything. There is no way to add a surface stress.

Often times we think of two-dimensional models as idealizing the evolution of the vertical average or something. So we can derive a two-dimensional momentum equation for example by taking the vertical average of the equation I wrote above, but over an entire dimension, say an ocean of depth $H$. In that case we get:

$$ \partial_t \bar u = - \frac{1}{H} \left ( \tau_z^\mathrm{top} - \tau_z^\mathrm{bottom} \right ) $$

where $H$ is the depth of the domain and $\bar u = H^{-1} \int_{-H}^0 u \, \mathrm{d} z$ is the vertical average of $u$. 

So one can think of "surface forcing" as being a body force acting on the vertically-averaged equations, which are integrated in $z$ and therefore two-dimensional in $x, y$. Is this the abstraction / physics that you are looking for @lgloege ?

But even in this case I strongly recommend to refer to the physical effect as a body force in 2D, rather than a "surface stress". It's not a surface stress, because a 2D model has no surface. We've developed the code intentionally to help users (but also people who read user's code) understand clearly the physics that are being applied. In particular, a 2D model has no grid spacing in the vertical direction (it's not easy to think of this as a model with just 1 grid point in the vertical direction, which is usually the way that people simplify 3D to 2D). So we can't apply a surface stress to the "top" grid cell, there is no $\Delta_z$.

In summary, surface or bottom boundary conditions _cannot be applied_ to models that are 2D in x, y. However, it is likely that the effect one is trying to achieve would use a body `Forcing` to apply a force in the interior of a 2D model.

**Comment by lgloege** [comment]
@glwagner Thank you so much for that detailed explanation. It makes perfect sense now why I can't apply a surface forcing. Instead of applying a body `Forcing` I just turned it into a 3D model with 1 vertical layer. This makes more sense in my head because then I can apply a true surface stress. 

```julia
using Pkg
Pkg.activate(".")

using Dates
using Random
using Printf

using Oceananigans
using Oceananigans.Units: minute, minutes, hour, hours, day, days, meter, meters, kilometer, kilometers

# --------------------------------------------------
# computing architecture
# --------------------------------------------------
architecture = CPU()

# --------------------------------------------------
# grid
# --------------------------------------------------
const Nx = 100  # number of x points 
const Ny = 100  # number of y points 
const Nz = 1    # number of z points 

const Lx = 100kilometers # (m) domain x extent
const Ly = 100kilometers # (m) domain y extent
const Lz = 10meters      # (m) domain depth

grid = RectilinearGrid(
    architecture, 
    topology = (Periodic, Periodic, Bounded),
    size = (Nx, Nx, Nz),
    halo = (3,3,2),
    x = (0, Lx),
    y = (0, Ly),
    z = (-Lz, 0)
)

# --------------------------------------------------
# coriolis
# --------------------------------------------------
coriolis = FPlane(f=1e-4)

# --------------------------------------------------
# closure
# --------------------------------------------------
closure = AnisotropicMinimumDissipation()

# --------------------------------------------------
# seawater buoyancy
# --------------------------------------------------
buoyancy = SeawaterBuoyancy(
    equation_of_state = LinearEquationOfState(
        thermal_expansion = 2e-4,
        haline_contraction = 8e-4)
)

# --------------------------------------------------
# surface stress
# --------------------------------------------------
u₁₀ = 10    # m s⁻¹, average wind velocity 10 meters above the ocean
ρₒ = 1026.0 # kg m⁻³, average density at the surface of the world ocean
cᴰ = 2.5e-3 # dimensionless drag coefficient
ρₐ = 1.225  # kg m⁻³, average density of air at sea-level
  
@inline wind_stress_x(x, y, t, p) = (- ρₐ / ρₒ * cᴰ * u₁₀ * abs(u₁₀)) * rand()
@inline wind_stress_y(x, y, t, p) = (- ρₐ / ρₒ * cᴰ * u₁₀ * abs(u₁₀)) * rand()

# --------------------------------------------------
# boundary conditions
# --------------------------------------------------
u_top_bcs = FluxBoundaryCondition(wind_stress_x, parameters=(k=4π, ω=3.0, τ=1e-4))
v_top_bcs = FluxBoundaryCondition(wind_stress_y, parameters=(k=4π, ω=3.0, τ=1e-4))

u_bcs = FieldBoundaryConditions(top = u_top_bcs) 
v_bcs = FieldBoundaryConditions(top = v_top_bcs)

T_bcs = FieldBoundaryConditions()
S_bcs = FieldBoundaryConditions()
c_bcs = FieldBoundaryConditions()

# --------------------------------------------------
# instantiate model
# --------------------------------------------------
model = NonhydrostaticModel(; grid, buoyancy,
                            timestepper = :QuasiAdamsBashforth2,
                            advection = WENO(),
                            tracers = (:T, :S, :c),
                            coriolis = coriolis,
                            closure = closure,
                            boundary_conditions = (u=u_bcs, v=v_bcs, T=T_bcs, S=S_bcs, c=c_bcs))


# --------------------------------------------------
# initial conditions
# --------------------------------------------------
const u_initial = 0.0
const v_initial = 0.0
const T_initial = 20.0
const S_initial = 35.0

# initial concentration value
# center indices (x, y) 
center_x_index = Int(grid.Nx / 2)  # center index in the x-direction
center_y_index = Int(grid.Ny / 2)  # center index in the y-direction
surface_z_index = grid.Nz          # index of surface z grid cell 

# initialize tracer values to zero
c_initial = model.tracers.c 

# grid cell volume
# note: this assumes grid cells all same size
dx = Lx / Nx       # (m) grid cell x length 
dy = Ly / Ny       # (m) grid cell y length
dz = Lz / Nz       # (m) grid cell z length
dv = dx * dy * dz  # (m3) grid cell volume

# set initial concentration to 1 mol / m3 at center grid cell
c_initial[center_x_index, center_y_index, surface_z_index] = 1

set!(model, u=u_initial, v=v_initial, T=T_initial, S=S_initial, c=c_initial)

# --------------------------------------------------
# simulation and outputs
# --------------------------------------------------
simulation = Simulation(model, Δt=10minutes, stop_time=365days)

wizard = TimeStepWizard(cfl=0.2)

simulation.callbacks[:wizard] = Callback(wizard, IterationInterval(10))


progress_message(sim) = @printf("Iteration: %04d, time: %s, Δt: %s, max(|u|) = %.1e ms⁻¹, wall time: %s\n",
                                iteration(sim), prettytime(sim), prettytime(sim.Δt),
                                maximum(abs, sim.model.velocities.u), prettytime(sim.run_wall_time))

add_callback!(simulation, progress_message, IterationInterval(20))


# --------------------------------------------------
# output writer
# --------------------------------------------------
filename = "output-tracer-release_$(string(today()))"

simulation.output_writers[:surface_slice_writer] = NetCDFOutputWriter(
    model, 
    (; model.velocities.u, model.velocities.v, model.tracers.S, model.tracers.T, model.tracers.c),
    filename = filename * ".nc",
    schedule = AveragedTimeInterval(1hour, window=1hour),
    indices=(:, :, grid.Nz),
    overwrite_existing = true
)

# --------------------------------------------------
# run simulation
# --------------------------------------------------
run!(simulation)
```
